torch_xla/experimental/pallas_kernels/quantized_matmul_kernel.py

```diff
@@ -14,69 +14,116 @@ def _quantize_array(
   x_ref_dtype = x_ref.dtype
   assert x_ref.shape == (batch_block_size,
                          in_block_size), "x_ref shape is not correct"
   assert w_ref.shape == (out_block_size,
                          in_block_size), "w_ref shape is not correct"
   assert scalar_ref.shape == (1,
                               out_block_size), "scalar_ref shape is not correct"
-  assert x_abs_max_val.shape == (
+  assert x_abs_max_ref.shape == (
       1, batch_block_size), "x_max_val shape is not correct"
   assert out_ref.shape == (batch_block_size,
                            out_block_size), "out_ref shape is not correct"
-  assert acc_ref.shape == (batch_block_size,
-                           out_block_size), "acc_ref shape is not correct"
-
-  @pl.when(in_idx == 0)
-  def _():
-    acc_ref[...] = jnp.zeros_like(acc_ref)
-
-  if quantize_activation:
-    x, x_scale = _quantize_array(x_ref[...], x_abs_max_val[...])
-    acc_ref[...] += jax.lax.dot_general(
-        x,
-        w_ref[...],
-        (((1,), (1,)), ((), ())),
-        preferred_element_type=jnp.int32,
-    )
+
+  if save_quant:
+    assert quantize_activation
+    assert q_x_scratch is not None
+    assert x_scale_scratch is not None
+    quant = out_idx == 0
   else:
-    acc_ref[...] += jax.lax.dot_general(
-        x_ref[...],
-        w_ref[...],
-        (((1,), (1,)), ((), ())),
-    )
-
-  @pl.when(in_idx == nsteps - 1)
-  def _():
-    acc = acc_ref[...]
-    scalar = scalar_ref[...]
-    acc *= scalar
+    assert q_x_scratch is None
+    assert x_scale_scratch is None
+    quant = quantize_activation
+
+  if save_acc:
+    assert acc_scratch is not None
+    is_first_step = in_idx == 0
+    is_last_step = in_idx == n_in - 1
+  else:
+    assert acc_scratch is None
+    is_first_step = True
+    is_last_step = True
+
+  def matmul_body(quant, is_first_step, is_last_step):
     if quantize_activation:
-      acc *= x_scale
-    out_ref[...] = acc.astype(x_ref_dtype)
+      if quant:
+        q_x_tmp, x_scale_tmp = _quantize_array(x_ref[...], x_abs_max_ref[...])
+        if save_quant:
+          q_x_scratch[...] = q_x_tmp
+          x_scale_scratch[...] = x_scale_tmp
+      else:
+        assert save_quant
+        q_x_tmp = q_x_scratch[...]
+        if is_last_step:
+          x_scale_tmp = x_scale_scratch[...]
+
+      acc = jax.lax.dot_general(
+          q_x_tmp,
+          w_ref[...],
+          (((1,), (1,)), ((), ())),
+          preferred_element_type=jnp.int32,
+      )
+    else:
+      acc = jax.lax.dot_general(
+          x_ref[...],
+          w_ref[...],
+          (((1,), (1,)), ((), ())),
+      )
+
+    if not is_first_step:
+      acc += acc_scratch[...]
+
+    if is_last_step:
+      acc *= scalar_ref[...]
+      if quantize_activation:
+        acc *= x_scale_tmp
+      out_ref[...] = acc.astype(x_ref_dtype)
+    else:
+      assert save_acc
+      acc_scratch[...] = acc
+
+  unfold_args((quant, is_first_step, is_last_step), (), matmul_body)
```

Comment:
could it be simplified? Since you have quant, is_first_step and is_last_step, can you directly call matmul_body(quant, is_first_step, is_last_step)?